pkg/providers/instanceprofile/instanceprofile.go

```diff
@@ -0,0 +1,131 @@
+	"github.com/aws/karpenter/pkg/apis/settings"
+	"github.com/aws/karpenter/pkg/apis/v1beta1"
+	awserrors "github.com/aws/karpenter/pkg/errors"
+)
+
+type Provider struct {
+	region string
+	iamapi iamiface.IAMAPI
+	cache  *cache.Cache
+}
+
+func NewProvider(region string, iamapi iamiface.IAMAPI, cache *cache.Cache) *Provider {
+	return &Provider{
+		region: region,
+		iamapi: iamapi,
+		cache:  cache,
+	}
+}
+
+func (p *Provider) Create(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) (string, error) {
+	tags := lo.Assign(nodeClass.Spec.Tags, map[string]string{
+		fmt.Sprintf("kubernetes.io/cluster/%s", settings.FromContext(ctx).ClusterName): "owned",
+		corev1beta1.ManagedByAnnotationKey:                                             settings.FromContext(ctx).ClusterName,
+		v1beta1.LabelNodeClass:                                                         nodeClass.Name,
+		v1.LabelTopologyRegion:                                                         p.region,
+	})
+	profileName := GetProfileName(ctx, p.region, nodeClass)
+
+	// An instance profile exists for this NodeClass
+	if _, ok := p.cache.Get(string(nodeClass.UID)); ok {
+		return profileName, nil
+	}
+	// Validate if the instance profile exists and has the correct role assigned to it
+	var instanceProfile *iam.InstanceProfile
+	out, err := p.iamapi.GetInstanceProfileWithContext(ctx, &iam.GetInstanceProfileInput{InstanceProfileName: aws.String(profileName)})
+	if err != nil {
+		if !awserrors.IsNotFound(err) {
+			return "", fmt.Errorf("getting instance profile %q, %w", profileName, err)
+		}
+		o, err := p.iamapi.CreateInstanceProfileWithContext(ctx, &iam.CreateInstanceProfileInput{
+			InstanceProfileName: aws.String(profileName),
+			Tags:                lo.MapToSlice(tags, func(k, v string) *iam.Tag { return &iam.Tag{Key: aws.String(k), Value: aws.String(v)} }),
+		})
+		if err != nil {
+			return "", fmt.Errorf("creating instance profile %q, %w", profileName, err)
+		}
+		instanceProfile = o.InstanceProfile
+	} else {
+		instanceProfile = out.InstanceProfile
+	}
+	if len(instanceProfile.Roles) == 1 {
+		if aws.StringValue(instanceProfile.Roles[0].RoleName) == nodeClass.Spec.Role {
+			return profileName, nil
+		}
+		if _, err = p.iamapi.RemoveRoleFromInstanceProfileWithContext(ctx, &iam.RemoveRoleFromInstanceProfileInput{
+			InstanceProfileName: aws.String(profileName),
+			RoleName:            instanceProfile.Roles[0].RoleName,
+		}); err != nil {
+			return "", fmt.Errorf("removing role %q for instance profile %q, %w", aws.StringValue(instanceProfile.Roles[0].RoleName), profileName, err)
+		}
+	}
+	if _, err = p.iamapi.AddRoleToInstanceProfileWithContext(ctx, &iam.AddRoleToInstanceProfileInput{
+		InstanceProfileName: aws.String(profileName),
+		RoleName:            aws.String(nodeClass.Spec.Role),
+	}); err != nil {
+		return "", fmt.Errorf("adding role %q to instance profile %q, %w", nodeClass.Spec.Role, profileName, err)
+	}
+	p.cache.SetDefault(string(nodeClass.UID), nil)
```

Comment:
i thought we were going to do something other than the UID here?

Reply:
UID makes sense for the cache, since we don't want previous versions of entries in the cache with the same nodeclass name to overlap